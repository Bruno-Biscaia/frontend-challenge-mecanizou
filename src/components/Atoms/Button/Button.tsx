import { cn } from '@/utils/cn';
import { ButtonHTMLAttributes, ReactNode } from 'react';

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  children: ReactNode;
  variant?: 'primary' | 'secondary' | 'outline' | 'ghost';
  size?: 'sm' | 'md' | 'lg';
  isLoading?: boolean;
}

export function Button({
  children,
  variant = 'primary',
  size = 'md',
  isLoading = false,
  disabled,
  className,
  ...props
}: ButtonProps) {
  const base =
    'inline-flex cursor-pointer items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none';

  const variants = {
    primary:
      'bg-indigo-600 text-white hover:bg-indigo-700 focus:ring-indigo-500',
    secondary:
      'bg-gray-100 text-gray-900 hover:bg-gray-200 focus:ring-gray-400',
    outline: 'border border-gray-300 text-gray-900 hover:bg-gray-50',
    ghost: 'text-gray-900 hover:bg-gray-100',
  };

  const sizes = {
    sm: 'px-3 py-1.5 text-sm',
    md: 'px-4 py-2 text-base',
    lg: 'px-5 py-3 text-lg',
  };

  return (
    <button
      disabled={isLoading || disabled}
      className={cn(base, variants[variant], sizes[size], className)}
      {...props}
    >
      {isLoading && (
        <svg
          className="animate-spin mr-2 h-4 w-4 text-white"
          xmlns="http://www.w3.org/2000/svg"
          fill="none"
          viewBox="0 0 24 24"
        >
          <circle
            className="opacity-25"
            cx="12"
            cy="12"
            r="10"
            stroke="currentColor"
            strokeWidth="4"
          />
          <path
            className="opacity-75"
            fill="currentColor"
            d="M4 12a8 8 0 018-8v4a4 4 0 00-4 4H4z"
          />
        </svg>
      )}
      {children}
    </button>
  );
}
