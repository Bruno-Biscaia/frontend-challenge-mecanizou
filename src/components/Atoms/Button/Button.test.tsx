import { render, screen, fireEvent } from '@testing-library/react';
import { Button } from './Button';

describe('Button component', () => {
  const onClick = jest.fn();

  it('renderiza o texto passado como children', () => {
    render(<Button onClick={onClick}>Comprar</Button>);
    expect(
      screen.getByRole('button', { name: /comprar/i })
    ).toBeInTheDocument();
  });

  it('chama a função onClick ao clicar', () => {
    render(<Button onClick={onClick}>Comprar</Button>);
    fireEvent.click(screen.getByRole('button', { name: /comprar/i }));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('mantém o mesmo snapshot', () => {
    const { container } = render(<Button onClick={onClick}>Comprar</Button>);
    expect(container).toMatchSnapshot();
  });
});
