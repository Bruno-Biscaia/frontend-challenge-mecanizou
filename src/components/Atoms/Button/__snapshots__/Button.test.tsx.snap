// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Button component mant√©m o mesmo snapshot 1`] = `
<div>
  <button
    class="inline-flex cursor-pointer items-center justify-center rounded-md font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none bg-indigo-600 text-white hover:bg-indigo-700 focus:ring-indigo-500 px-4 py-2 text-base"
  >
    Comprar
  </button>
</div>
`;
